hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/clustering/run/strategy/MultipleSparkJobExecutionStrategy.java

```diff
@@ -361,7 +369,15 @@ private HoodieData<HoodieRecord<T>> readRecordsForGroupBaseFiles(JavaSparkContex
           List<Iterator<HoodieRecord<T>>> iteratorsForPartition = new ArrayList<>();
           clusteringOpsPartition.forEachRemaining(clusteringOp -> {
             try {
-              Schema readerSchema = HoodieAvroUtils.addMetadataFields(new Schema.Parser().parse(writeConfig.getSchema()));
+              TableSchemaResolver schemaUtil = new TableSchemaResolver(getHoodieTable().getMetaClient());
+              Schema readerSchema;
+              try {
+                readerSchema = schemaUtil.getTableAvroSchema(true);
+              } catch (Exception e) {
+                LOG.warn(e.getMessage());
+                readerSchema = HoodieAvroUtils.addMetadataFields(new Schema.Parser().parse(writeConfig.getSchema()));
+              }
```

Comment:
So we are using the latest table schema for clustering, that makes sense to me.

Reply:
We actually got similiar fix for compaction, in file `HoodieCompactor.java`:

```java
    // Here we firstly use the table schema as the reader schema to read
    // log file.That is because in the case of MergeInto, the config.getSchema may not
    // the same with the table schema.
    try {
      if (StringUtils.isNullOrEmpty(config.getInternalSchema())) {
        Schema readerSchema = schemaResolver.getTableAvroSchema(false);
        config.setSchema(readerSchema.toString());
      }
    } catch (Exception e) {
      // If there is no commit in the table, just ignore the exception.
    }
```

Reply:
Yes, let's first obtain schema information from the hudi table for clustering. If an exception occurs, we will use schema configuration as a backup

Reply:
Okay, let me take a look and see if we can make the necessary changes together

Reply:
I think the fix in this path looks fine, can we just abstract the schema fething logic into a separate method that maybe called `getLatestTableSchema`, I mean move the try-catch and fallback caluse into it.